test/parallel_api/numeric/numeric.ops/in_place_scan.pass.cpp

```diff
@@ -19,64 +19,532 @@
+        auto new_policy2 = make_new_policy<new_kernel_name<Policy, 1>>(exec);
+        /*auto res2 = */oneapi::dpl::exclusive_scan(new_policy2, keys_first, keys_last, keys_first, kInitValue, DefaultBinaryOp<KeyT>());
+        exec.queue().wait_and_throw();
+
+        retrieve_data(host_keys);
+        check_values(source_host_keys_state, host_keys.get(), n, DefaultBinaryOp<KeyT>());
+
+        // call algorithm with equality comparator
+        initialize_data(host_keys.get(), n);
+        update_data(host_keys);
+
+        auto new_policy3 = make_new_policy<new_kernel_name<Policy, 2>>(exec);
+        /*auto res3 = */oneapi::dpl::exclusive_scan(new_policy3, keys_first, keys_last, keys_first, kInitValue, BinaryOperation());
+        exec.queue().wait_and_throw();
+
+        retrieve_data(host_keys);
+        check_values(source_host_keys_state, host_keys.get(), n, BinaryOperation());
+    }
+#endif
+
+    // specialization for host execution policies
+    template <typename Policy, typename Iterator1, typename Size>
+    typename ::std::enable_if<
+#if TEST_DPCPP_BACKEND_PRESENT
+        !oneapi::dpl::__internal::__is_hetero_execution_policy<typename ::std::decay<Policy>::type>::value &&
+#endif
+            is_base_of_iterator_category<::std::random_access_iterator_tag, Iterator1>::value,
+        void>::type
+    operator()(Policy&& exec, Iterator1 keys_first, Iterator1 keys_last, Size n)
+    {
+        typedef typename ::std::iterator_traits<Iterator1>::value_type KeyT;
+
+        // call algorithm with no optional arguments
+        initialize_data(keys_first, n);
+
+        // copy source host keys state into temp buf
+        typedef typename ::std::iterator_traits<Iterator1>::value_type ValT;
+        const std::vector<ValT> source_host_keys_state(keys_first, keys_first + n);
+
+        auto res1 = oneapi::dpl::exclusive_scan(exec, keys_first, keys_last, keys_first, kInitValue);
+        check_values(source_host_keys_state, keys_first, n);
+
+        // call algorithm with equality comparator
+        initialize_data(keys_first, n);
+        auto res2 = oneapi::dpl::exclusive_scan(exec, keys_first, keys_last, keys_first, DefaultBinaryOp<KeyT>());
+        check_values(source_host_keys_state, keys_first, n, DefaultBinaryOp<KeyT>());
+
+        // call algorithm with addition operator
+        initialize_data(keys_first, n);
+        auto res3 = oneapi::dpl::exclusive_scan(exec, keys_first, keys_last, keys_first, kInitValue, BinaryOperation());
+        check_values(source_host_keys_state, keys_first, n, BinaryOperation());
+    }
+
+    // specialization for non-random_access iterators
+    template <typename Policy, typename Iterator1, typename Size>
+    typename ::std::enable_if<!is_base_of_iterator_category<::std::random_access_iterator_tag, Iterator1>::value, void>::type
+    operator()(Policy&& exec, Iterator1 keys_first, Iterator1 keys_last, Size n)
+    {
+    }
+};
+
+DEFINE_TEST_1(test_exclusive_scan_non_inplace, BinaryOperation)
```

Comment:
4 code block are very similar...
M.b. we can keep just 1 such code block?

Reply:
Partially fixed, At whole there are 4 test and 4 blocks of code, sometimes duplicated.